include/ttmlir/Dialect/TTNN/IR/TTNNOps.td

```diff
@@ -755,6 +755,23 @@ def TTNN_MatmulOp : TTNN_NamedDPSOp<"matmul"> {
 }
 // ANCHOR_END: adding_an_op_matmul_ttnn
 
+def TTNN_BroadcastOp : TTNN_NamedDPSOp<"broadcast"> {
```

Comment:
@sdjordjevicTT The TTNN team has confirmed to me that there is no specific plan to add broadcastOp in TTNN. They only plan to support implicit broadcast operations. I originally opened this issue https://github.com/tenstorrent/tt-metal/issues/13566 in order to support broadcast for all operands and now they have multiple open issues tracking support for it e.g. https://github.com/tenstorrent/tt-metal/issues/13643. I am not yet sure if we want to request an explicit BroadcastOp since we can always add a workaround to handle them, until such time we see a case where we cannot.

However they are still actively working on it and it seems that it might take some time until broadcastOp support is added to all Ops. As for a list of Ops that support implicit broadcast, it is ever changing and there is no official tracker. Since we want to bring up model support that need Broadcast, so I created this workaround to get us going until support is being added by TTNN team and add exclusions when we hit cases that don't have implicit broadcast support yet. Therefore it is still a workaround rather than a transform pass as it is temporary solution.

Reply:
I have updated the PR to extend the exclusion for reshape op to all unary ops.

Reply:
This workaround is definitely not a long term solution, so I opened this issue with tt-metal to request for an explicit BroadcastOp or else an alternative. 
https://github.com/tenstorrent/tt-metal/issues/16015

Reply:
Did you take a look at `ttnn.repeat` https://docs.tenstorrent.com/tt-metal/latest/ttnn/ttnn/api/ttnn.repeat.html#ttnn.repeat (disregard deceptive documentation, `repeat_dims` is `Shape` of the same size as a rank of an input tensor)? Any broadcast is just a special case of repeat. I still got some of those comments from your PR, so in my opinion the best course of action is:
- Adding `ttir.repeat` and `ttnn.repeat` (with runtime support).
- Lowering `ttir.broadcast` to `ttnn.repeat`.
- We already have the `TTIR_Broadcastable` interface in the TTIR dialect, we also need to add the `Commutative` trait and mark commutative ops with it. Only with those traits in place we can make optimizations like "if second operand is broadcastable into the shape of first operand, reverse their order", and some other foldings that you've done or suggested here.
- Track the progress on the metal side, when both operands can be implicitly broadcastable do even more folding.

Another alternative that I've looked into before I came across `ttnn.repeat` was `ttnn.concat` (with fastpow this is also a sound idea for decomposition), but that would almost certainly have a higher performance penalty and would be harder for implementation. The advantage is that it's already implemented op in dialects, but as I said there is a much higher risk of perf penalty, so I would only test that option after we have `ttnn.repeat` (more out of curiosity at that moment).

I would close this PR, as adding a non-TNN ops to a TTNN dialect should be a last resort.

Reply:
This is actually a good example, because it's still doable with reshape(1, 3, 1), followed by repeat(2, 1, 2), but if ttnn implicit broadcast works like a numpy implicit broadcast (it seems that's the case from a quick check) then it won't be possible with implicit broadcast as 3 != 2 and none of them are 1.
Perf-wise I'm not sure we are gaining or loosing anything, we would have to confirm if they are taking any advantage of implicit broadcast on the metal side before we jump to the conclusion.

Reply:
Yes I meant about the cases where we can legally perform implicit broadcast. Metal team has confirmed that implicit broadcast performs _much better_ than explicit ops, so I think we should design to use it as much as possible.

Now I guess the main question is whether we should still ask for an explicit broadcastOp or use a combination of `reshape` followed by `repeat`. For functionality purpose, I suppose it doesn't matter, but makes it for a simpler design if we can have a single op to do both.

Reply:
The second design question is for cases where do not (yet) have implicit broadcast support, we have two options: 

1) Use the workaround added by this PR to insert an AddOp with zero tensor to carry the implicit broadcast. 
OR
2) Insert a repeat op to perform the broadcast operation. 

In both cases we will add an extra instruction. However I think (1) is better because it gives us the option to remove the workaround as soon as metal team is done with adding support for implicit broadcast to all binary and ternary eltwise cases (probably ready by the time we start optimizing for performance).

Reply:
Let's break down all the cases of broadcast we have:
1. Broadcast output is used as an operand to ops which support implicit broadcasting
2. Broadcast output is used as an operand to ops which will support implicit broadcasting in the future
3. Broadcast output is used as an operand to ops which will never support implicit broadcasting (for example unary ops or returning broadcast output as the final result of the graph)

1st case should be handled by the TTIR->TTIR fusion pass, fusing `ttir.broadcast + ttir.other_op -> ttir.other_op`. 2nd case will in the future be handled the same way as the 1st case, but for now we need a workaround. 3rd case needs a long term solution.

So now let's talk about potential solutions:
- @uazizTT provided a solution with the AddOP. This can be used only for the broadcasts that fall into 2nd case.
- @azecevicTT provided another solution with using RepeatOp. This can be used for both 2nd case and 3rd case, only drawback is that it sometimes requires ReshapeOp before RepeatOp.
- Asking for Metal support of explicit broadcast op in ttnn.

Deciding on these solutions depends on informations from the Metal team, are they fundamentally against **ever** adding an explicit broadcast op, or is this just not on their radar right now and they **might** add it in the future? We should first communicate this with them and get their feedback.

If they are not fundamentally against **ever** adding an explicit broadcast op but they just don't have it on their radar right now, then this seems like the best way to proceed:
- Add `ttnn.broadcast` op and lower unfused `ttir.broadcast` ops into that.
- Add a workaround decomposition pattern to decompose `ttnn.broadcast` into `ttnn.repeat` or `ttnn.reshape + ttnn.repeat` if reshape is needed. For further optimization for broadcasts that fall into 2nd case above we can use the solution that Usman provided and decompose into `ttnn.zeros + ttnn.add` instead, but I am not sure if it will actually have a better perf and if it is worth complicating the pattern with two different decompositions. I think we need to test the perf and decide based on that.
- Open issue to Metal team to add explicit broadcast op consistent with `torch.broadcast_to`, and keep it low priority for now because we have workarounds. If we confirm in the future that those workarounds are having a big perf impact in our models, raise the priority to the Metal team.

Otherwise:
- Decompose unfused `ttir.broadcast` ops similar to workaround decomposition described above, but now as part of TTIR->TTIR pass or TTIR->TTNN pass, I am not sure what is the best place because it depends on the order of Decomposition/Fusing passes.
- Once Metal adds support for implicit broadcast for other ops, then we will just change the fusing conditions so broadcasts that fall into 2nd case above will also be fused and won't be decomposed.

What others think?

Reply:
@mrakitaTT why does the WA that @uazizTT proposed not suitable for 3? As far as I see it, it's just a NOP to get the desired output shape? 

That said, the `ttnn.repeat` proposal from @azecevicTT seems like it's pretty much what we want, an explicit broadcast op, with just slightly different semantics, that is, we need to explicitly unsqueeze the tensor if the rank is different. I don't mind the unsqueeze since it makes the graph easier to read, and if not on a tile dim, will be free on device. 

My vote is that we lower to `ttnn.repeat` when we can't fuse the broadcast in ttir.

Reply:
I think we all are converging to the same direction, so here is what I propose to do next:

1) For case (3) above, lower `ttir.broadcast` to `ttnn.repeat ` and keep `ttir.reshape` separate. Once Metal adds support for explicit broadcastOp, then the `ttnn.reshape `and `ttnn.repeat` can be folded to `ttnn.broadcast` as an optimization. 
2) Keep the above workaround for rest of the cases i.e. for binary and ternary eltwise ops where we expect to have imminent support for implicit broadcast. This workaround is supposed to be temporary and removed, but until then we keep `ttnn.broadcast`, which will be removed at the same time when this workaround is removed.

Reply:
I guess my question is what would `ttnn.broadcast` do that `ttnn.repeat` does not? Just the implicit rank changes?

Reply:
> I guess my question is what would `ttnn.broadcast` do that `ttnn.repeat` does not? Just the implicit rank changes?

Yes that's right, so it can handle the case below in a single instruction, we still need two instructions to do the same in the above proposals. 

```
%result = "stablehlo.broadcast_in_dim"(%operand) {
  broadcast_dimensions = array<i64: 2, 1>
} : (tensor<1x3xi32>) -> tensor<2x3x2xi32>
```

Reply:
OK, in that case, in my opinion, this is a case where we have a small API difference between two stacks, and it doesn't really warrant a new op to completely align them. That's why we have a compiler 🙂 

Also, when you say

> Keep the above workaround for rest of the cases 

Do you mean the WA using add with zeros? If we're adding a path to lower broadcast to `ttnn.repeat` I would rather just use that in cases where we can't fuse.

Reply:
Chatted offline with @uazizTT, summarizing here for others' context. He explained that he wanted to keep the WA for ops that should support broadcast very obvious and remove it quickly since apparently all eltwise binary ops will have support for implicit broadcast on either argument by end of January, so his preference was to keep the WA different than the path for lowering broadcasts that will most likely never be supported (i.e. unary). 

I'm somewhat indifferent as to which WA we use, so I'll leave it to others to chime in if they have a strong opinion.

Reply:
> why does the WA that @uazizTT proposed not suitable for 3? As far as I see it, it's just a NOP to get the desired output shape?

@AleksKnezevic It can't handle broadcast shapes like [this](https://github.com/tenstorrent/tt-mlir/pull/1542#discussion_r1886932865) one, implicit broadcasting has [strict rules](https://pytorch.org/docs/stable/notes/broadcasting.html) when it can be applied. I forgot to mention it in definition of 3rd case, updated now.

That's why we need to have decomposition to `ttnn.reshape + ttnn.repeat`. As I've said we could also check if broadcast is using shapes which can be implicitly broadcasted and in that case use the solution that Usman provided and decompose into `ttnn.zeros + ttnn.add` instead, but I am not sure if it will actually have a better perf and if it is worth complicating the pattern with two different decompositions. I think we would need to do the perf tests and decide based on that.

Reply:
> OK, in that case, in my opinion, this is a case where we have a small API difference between two stacks, and it doesn't really warrant a new op to completely align them. That's why we have a compiler 🙂

I'm not sure, torch has both: [torch.repeat](https://pytorch.org/docs/stable/generated/torch.Tensor.repeat.html) and [torch.broadcast_to](https://pytorch.org/docs/stable/generated/torch.broadcast_to.html). I don't see why ttnn wouldn't have both. But as I've said this is something we should probably discuss with them before making a decision if this will be a workaround or long term decomposition.

Reply:
@uazizTT Can you point me to discussion where metal team confirmed that implicit broadcast will be more performant than explicit repeat? If you are referring to https://github.com/tenstorrent/tt-metal/issues/13643, even though issue description isn't clear enough, I think it's a different case than what we have here. I'm far from an expert in this matter, but as far as my understanding goes (and I urge someone to correct me, as most of these are just my assumptions which I didn't verify), this is the reason why implicit broadcast is more performant than repeat in that issue. For simplicity I will denote tensor<64x64> as small, and tensor<4096x64x64> as big, I will do just a dimensional analysis:

(1)
materialized_big = ttnn.repeat(small)
materialized_big = ttnn.binary(materialized_big, big)

vs

(2)
materialized_big = ttnn.binary(small, big)

In (1) we have two materializations of big, that's why it's expected to be slower than (2). What we have here is:

(3)
materialized_big = ttnn.repeat(small)

vs

(4)
materialized_big = ttnn.add(small, big)

Here we have one materialization in both (3) and (4), with added eltwise operation in (4), so I would consider it a bug on the metal side if (4) is ever performing significantly better than (3). That's why I would like to see a direct comparison of these two cases and try to analyze the results and their implication before we jump to a conclusion, as these results could have far-reaching consequences.

@mrakitaTT They have a very good reason for having both `repeat` and `broadcast_to`, because they have a concept of view. As far as I am aware `ttnn` doesn't have a concept of view, that's why `broadcast` would be almost redundant (with the exception of possible `reshape` before `repeat`). I'm not against implementing in on the`ttnn` side as a convenience feature, but I would consider it a very low priority until we have (if ever) something that resembles the concept of view.

Ideally implicit broadcast should be a convenience feature, not a performance enhancer, alas I can see how that can be a case in (1) vs (2). That doesn't mean that a lot of other ops wouldn't benefit from the same principle (eltwise unary from the top of my head). This whole problem is rooted in the lack of concept of view and eager materialization of tensor, so lazy materialization would be something to think about on the compiler's side in the future.

Reply:
> @uazizTT Can you point me to discussion where metal team confirmed that implicit broadcast will be more performant than explicit repeat? 

It was a Slack discussion, not sure if you can open [this link](https://tenstorrent.slack.com/archives/C0853T28RPE/p1734363820696949?thread_ts=1734363491.686539&cid=C0853T28RPE), otherwise posting it here too for reference:

```
Usman Aziz:
Good Morning everyone.
I have a question about performance for implicit-broadcast cases, i.e. consider that case where we need implicit broadcast for any operand of AddOp, is the performance penalty less to use implicit-broadcast or the same if we add an explicit broadcastOp before the AddOp?

Eyon:
It will be much better to use broadcast within the add operation because you save on needing to dispatch.   Anytime you can avoid making a call to run an operation across the PCI bus you will save. (edited)
```

Just to clarify, I think it will only be beneficial for performance if there an existing Eltwise op to carry the implicit broadcast, which we normally do in case of `Binary` or `Ternary` eltwise ops, otherwise using a `ttnn.repeat` is better than having to add a new eltwise op just to carry the implicit broadcast e.g. in `Unary` ops.

Reply:
Yeah, that's what I expected, it's a (1) vs (2) case from my previous post. I believe that we should go with `repeat` as the most general case. When `repeat` has the same semantics as `broadcast` additional `reshape` should act as `unsqueeze`, I don't know how it's implemented in `ttnn` but hopefully it's noop as far as data is concerned, because only metadata changes. For binary ops that support implicit broadcasting we should fold explicit broadcast as much as we can for performance reasons. To not repeat myself, https://github.com/tenstorrent/tt-mlir/pull/1542#discussion_r1884618371. The only difference from that comment is that I would add (at this moment) a `TTNN_ImplicitlyBroadcastable` interface, to track all of the ops that are implicitly broadcastable, and try to fold `ttnn.reshape` followed `ttnn.repeat` whose only user is `TTNN_ImplicitlyBroadcastable` op.
For generality, I think `repeat` will be more than enough, after that we can concentrate on these foldings, at that moment situation with implicitly broadcastable ops will be clearer, so we can make further decisions, and `repeat` leaves us room for those decisions.

Reply:
Great discussion, I am trying to summarize things. At this moment, we want to optimize for generality, hence based on all inputs, we have the following next steps:
1. Close this PR, as we don't want to work around this, we want a proper and long-term solution
2. Introduce Repeat op in both TTIR and TTNN dialects:
```mlir
%0 = "ttir.repeat"(%arg0) { repeatDims = {...
%0 = "ttnn.repeat"(%arg0) { repeatDims = {...
```
3. Lower TTIR broadcast to TTNN_RepeatOp during TTIR to TTNN conversion pass:
  - Lower TTIR_BroadcastOp 1:1 to TTNN_RepeatOp
    - Example: `1x64x64 -> 64x64x64: repeat(64, 1, 1)`
  - Combination of reshape + repeat
    - Example: `64x64 -> 64x64x64: reshape(1, 64, 64) + repeat(64, 1, 1)`

This will cover 100% of generality, and then we can start optimizing for performance:
1. Fold all implicitly broadcastable operations

This approach ensures long-term generality, avoids adding unnecessary ops, and relies on compiler transformations to optimize broadcast handling.

Reply:
Thanks @sdjordjevicTT for the summary, I agree with this conclusion that we should lower to repeat to get a working solution. I wanted to clarify for completeness so we have all discussion in one place:

The broadcast op will only be used for `unary`, `binary` or `ternary` eltwise ops. There is no other ops that we see have broadcast so the above becomes our general case. That simplifies things because as a long term solution, we never fold for unary and broadcast will always be lowered to repeat. Hence, implicit broadcast will only become a property for binary and ternary Eltwise Ops. 

Now for optimization, we can either:
1) Keep track of the Ops that can implicitly fold and keep on amending the list as metal adds support
OR
2) Make it a rule that all binary and ternary ops can always fold and use a workaround to handle the cases where metal support is incomplete e.g. Ops that can only implicit broadcast some operands. 

I think both of the above solutions can work in the optimization phase.